add multibyte public version support (zcash support)

D-Nice opened the pull request:
- adds support for coins, such as zcash, that use two or more bytes for
versioning of public addresses. This was done by converting all versions
passed as Buffers to coinstring, which the library insists on for
multibyte support.
- added zcash into test cases, required adding a conditional whereby
versions are passed into `fromWif` in the case the coin doesn't follow
the previous convention of priv - pub = 0x80. Can be easily done by also
leveraging coininfo.

Should result in all coins currently listed on coininfo being supported, except for Decred, which uses multibytes for the private key. Had an implementation for that too, but in general it seems that private keys are not intended to be supported with Decred for whatever reason.

D-Nice committed:
add multibyte public version support
- adds support for coins, such as zcash, that use two or more bytes for
versioning of public addresses. This was done by converting all versions
passed as Buffers to coinstring, which the library insists on for
multibyte support.
- added zcash into test cases, required adding a conditional whereby
versions are passed into `fromWif` in the case the coin doesn't follow
the previous convention of priv - pub = 0x80. Can be easily done by also
leveraging coininfo.

D-Nice commented:
Yes, this is not something I've looked at, I've presumed big-endian as that's what Zcash appears to use. Do you know of any that may use LE? However, can the library not just expect all to be BE, and even if LE, for it to be pre-converted before being passed, into an integer that in BE form will correspond correctly?

D-Nice commented:
As a note, I've had a look at various coins, and the only one other than zcash copypastas I've seen using multibytes is decred, which I've been able to find no real deep documentation/protocol info on, and has an additional complication of having the private version be multibyte as well (which I do have working but didn't find useful to push).

This is a note on the endianness specified in the Zcash Protocol Specification: `All integers in Zcash-specific encodings are unsigned, have a fixed bit length, and are encoded in little-endian byte
order unless otherwise specified.` However, the version is apparently specified in the paper as being 2 distinct 8-bit values (from my understanding), rather than a single 16-bit uint, in which case the endianness does not matter. Meaning the current bufferizeVersion method is the way to go, as I'd expect the hex bytes as specified to be passed, and it works seamlessly with the current definitions in `coininfo`.

If you can recommend what would be needed to get this merged, let me know. My other option is just using a forked version, as this is a dependency for one of my own projects, but this being upstream is preferable and beneficial to others, if you guys deem it sufficient.

fanatid commented on lib/util.js:
```diff
@@ -23,8 +23,18 @@ function normalizeVersions (versions) {
+function bufferizeVersion (version) {
+  var length
+  if (version === 0 || version === 1) length = 1
+  else if (version > 1) length = Math.ceil((Math.log(version) / Math.log(2)) / 8)
```
What if version will be less than 256, but keys should be encoded with 2 bytes?

fanatid replied:
Better push new code to this PR.

D-Nice committed:
add support for pub version with leading zero byte(s)
- fixes previously incorrect length formula in `beUIntToBuffer`. Was
off-by-one for %256.
- stricter type checking on bufferizeVersion. Can handle BE uint, or in
case of LE uint or hex bytes with leading zeroes, a direct hex string is
expected to be provided.
- added fake multibytetestcoin, sharing same private byte as bitcoin,
but containing public version with a leading 00 byte. Public prefix
should result in `1337D` (chosen by coincidence using bs58check :/, but
since first char would be 1 due to 00, it seemed to look
nicest/different/obvious).

D-Nice commented:
New push.

Fixes aforementioned length issue.

Handles LE or leading zero instances by requiring a hex string sequence instead of passing a uint for the version. 

Please let me know if this is an acceptable change, and/or of other issues.

fanatid commented on lib/util.js:
```diff
@@ -24,11 +24,37 @@ function normalizeVersions (versions) {
+    // expects BE uint
+    return beUIntToBuffer(version)
+  }
+}
```
```js
if (typeof version === 'string') return hexStringToBuffer(version)
if (typeof version === 'number') return beUIntToBuffer(version)
throw new Error('invalid version type.')
```

D-Nice replied:
Good call on this :+1: will update with that

fanatid commented on lib/util.js:
```diff
@@ -24,11 +24,37 @@ function normalizeVersions (versions) {
+  if (!isValidHexString(input)) throw new Error('invalid hex string.')
+  // omit 0x for buffer
+  var re = /([\dA-Fa-f]{2})+$/g
+  const sanitizedInput = input.match(re)[0]
```
You do not need regexp for removing `0x`.

```js
var isValidRE = /^(0x)?([\dA-Fa-f]{2})+$/g
if (!isValidRE.text(input)) throw new Error('invalid hex string.')
return new Buffer(input.slice(2), 'hex')
```

D-Nice replied:
Used regexp to provide the option of providing the hex sequence as either "0xdeadbeef" or "deadbeef". If you feel "0x" prefix should be forced, then that change could be made.

fanatid replied:
```js
return new Buffer(input.slice(input.slice(0, 2) == '0x' ? 2 : 0), 'hex')
```
:)

D-Nice replied:
I will change to that :). Just wondering for the reason to omit using regexp, as I tend to use it when the opportunity presents itself for formatting/sanitizing user input, based on years old recommendations?

D-Nice committed:
improves new utils codebase
- implementing @fanatid recommended changes
- streamlining new utils function codebase, changing hex sequence
parsing from regexp to slicing.
- added non 0x prefixed multibytetestcoin version into fixtures.

RyanZim commented:
This is semver-minor, right?

fanatid commented:
Yes, I think this is semver-minor.

RyanZim commented:
Didn't realize this package was so out of date; published as 3.0.0 along with dropping old Node version support.
